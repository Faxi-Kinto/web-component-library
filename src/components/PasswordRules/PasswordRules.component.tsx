import React from 'react';
import * as Styled from './PasswordRules.styles';
import { theme } from '../../config';
import Icon from '../Icon';
import Text from '../Text';
import validatorsRegexes from '../../helpers/validationRegexes';
import { pxToRem } from '../../styles/basics';

export const hasNumber = (password: string) =>
  validatorsRegexes.hasNumber.test(password);
export const hasLowerCaseLetter = (password: string) =>
  validatorsRegexes.passwordLowerCase.test(password);
export const hasUpperCaseLetter = (password: string) =>
  validatorsRegexes.passwordUpperCase.test(password);

/**
 * @name PasswordRules
 *
 * PasswordRules molecule renders error messages based on password string
 *
 * @returns {JSX}
 */

export type PasswordRulesProps = {
  password: string;
};

const PasswordRules: React.FC<PasswordRulesProps> = (
  props: PasswordRulesProps
): JSX.Element => {
  const { password } = props;
  const requiredLength = password.length >= 6;
  const hasLowAndUpperCase =
    hasLowerCaseLetter(password) && hasUpperCaseLetter(password);

  return (
    <Styled.Container className="password_rules">
      <div className="password_rules__rule">
        <div className="password_rules__rule__icon">
          {renderValidationStatus(requiredLength)}
        </div>
        <div>
          <Text.Body>Contains at least 6 characters</Text.Body>
        </div>
      </div>
      <div className="password_rules__rule">
        <div className="password_rules__rule__icon">
          {renderValidationStatus(hasNumber(password))}
        </div>
        <div className="password_rules_wrapper_password_rule_number--message">
          <Text.Body>Contains at least one number (0-9)</Text.Body>
        </div>
      </div>
      <div className="password_rules__rule">
        <div className="password_rules__rule__icon">
          {renderValidationStatus(hasLowAndUpperCase)}
        </div>
        <Text.Body>
          Contains both lower (a-z) and upper case letters (A-Z)
        </Text.Body>
      </div>
    </Styled.Container>
  );

  function renderValidationStatus(validStatus: boolean) {
    if (validStatus) {
      return (
        <Icon
          name="correct"
          color={theme.pallet.TURQUOISE}
          size={pxToRem('12px')}
        />
      );
    } else {
      return (
        <Icon
          name="incorrect"
          color={theme.pallet.ERROR}
          size={pxToRem('14px')}
        />
      );
    }
  }
};

export default PasswordRules as React.FC<PasswordRulesProps>;
