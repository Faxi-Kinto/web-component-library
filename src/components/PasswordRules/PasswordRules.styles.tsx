import styled from 'styled-components';
import { pxToRem } from '../../styles/basics';
import { flexRow } from '../../styles/flex';
import { marginChildren } from '../../styles/spacings';
import { size } from '../../styles/basics';

export const Container = styled.div`
  width: 100%;
  ${marginChildren(`0 0 ${pxToRem('10px')}`)};

  .password_rules {
    &__rule {
      display: flex;
      ${flexRow('flex-start', 'flex-start')};
      ${marginChildren(`0 ${pxToRem('10px')} 0 0`)};

      &__icon {
        margin-top: ${pxToRem('2px')};
        ${size(pxToRem('10px'))};
      }
    }
  }
`;
