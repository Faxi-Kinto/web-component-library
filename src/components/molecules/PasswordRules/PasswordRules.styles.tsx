import styled from 'styled-components';
import { pxToRem } from 'styles/fonts';

export const Container = styled.div`
  width: 100%;

  .password_rules_wrapper {
    &_password_rule_characters {
      display: flex;
      &--icons {
        display: inline-block;
        width: ${pxToRem('26px')};
        div {
          display: inline-block;
          margin-right: 10px;
        }
      }

      &--message {
        display: inline-block;
        p {
          display: inline-block;
        }
        span {
          font-weight: 500;
        }
      }
    }

    &_password_rule_number {
      display: flex;
      &--icons {
        display: inline-block;
        width: ${pxToRem('26px')};
        div {
          display: inline-block;
          margin-right: 10px;
        }
      }

      &--message {
        display: inline-block;
        p {
          display: inline-block;
        }
        span {
          font-weight: 500;
        }
      }
    }

    &_password_rule_letters {
      display: flex;
      &--icons {
        display: inline-block;
        width: ${pxToRem('26px')};
        div {
          display: inline-block;
          margin-right: 10px;
        }
      }

      &--message {
        display: inline-block;
        p {
          display: inline-block;
        }
        span {
          font-weight: 500;
        }
      }
    }
  }
`;
