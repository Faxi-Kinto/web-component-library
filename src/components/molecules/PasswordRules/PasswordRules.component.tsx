import React from 'react';
import * as Styled from './PasswordRules.styles';
import { Icon } from 'components';
import { theme } from 'config';
import { Text } from 'components';
import validatorsRegexes from 'store/FormProvider/validationRegexes';

export const hasNumber = (password: string) =>
  validatorsRegexes.hasNumber.test(password);
export const hasLowerCaseLetter = (password: string) =>
  validatorsRegexes.passwordLowerCase.test(password);
export const hasUpperCaseLetter = (password: string) =>
  validatorsRegexes.passwordUpperCase.test(password);

/**
 * @name PasswordRules
 *
 * PasswordRules molecule renders error messages based on password string
 *
 * @returns {JSX}
 */

export type PasswordRulesProps = {
  password: string;
};

const PasswordRules: React.FC<PasswordRulesProps> = (
  props: PasswordRulesProps
): JSX.Element => {
  const { password } = props;
  const requiredLength = password.length >= 6;
  const hasLowAndUpperCase =
    hasLowerCaseLetter(password) && hasUpperCaseLetter(password);

  return (
    <Styled.Container>
      <div className="password_rules_wrapper">
        <div className="password_rules_wrapper_password_rule_characters">
          <div className="password_rules_wrapper_password_rule_characters--icons">
            {renderValidationStatus(requiredLength)}
          </div>
          <div className="password_rules_wrapper_password_rule_characters--message">
            <Text.Body>Contains at least 6 characters</Text.Body>
          </div>
        </div>
        <div className="password_rules_wrapper_password_rule_number">
          <div className="password_rules_wrapper_password_rule_number--icons">
            {renderValidationStatus(hasNumber(password))}
          </div>
          <div className="password_rules_wrapper_password_rule_number--message">
            <Text.Body>Contains at least one number (0-9)</Text.Body>
          </div>
        </div>
        <div className="password_rules_wrapper_password_rule_letters">
          <div className="password_rules_wrapper_password_rule_letters--icons">
            {renderValidationStatus(hasLowAndUpperCase)}
          </div>
          <div className="password_rules_wrapper_password_rule_letters--message">
            <Text.Body>
              Contains both lower (a-z) and upper case letters (A-Z)
            </Text.Body>
          </div>
        </div>
      </div>
    </Styled.Container>
  );

  function renderValidationStatus(validStatus: boolean) {
    if (validStatus) {
      return <Icon name="correct" color={theme.pallet.TURQUOISE} />;
    } else {
      return <Icon name="incorrect" color={theme.pallet.ERROR} />;
    }
  }
};

export default PasswordRules as React.FC<PasswordRulesProps>;
