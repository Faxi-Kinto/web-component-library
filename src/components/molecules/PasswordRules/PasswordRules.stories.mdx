import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";

import PasswordRules from "./";
import { withKnobs, text } from '@storybook/addon-knobs';

<Meta title="Core|Molecules/PasswordRules" />

# PasswordRules Component

It contains an icon and error message.

Based on the received password string a correct or incorrect icon is displayed.

## Implementation

#### Default
<Preview>
  <Story name="default" decorators={[withKnobs]}>
    <PasswordRules password={text('Password: ', 'Password1')} />
  </Story>
</Preview>

#### With Error
<Preview>
  <Story name="With an error">
    <PasswordRules password="" />
  </Story>
</Preview>

## PasswordRules Props

<Props of={PasswordRules} />
