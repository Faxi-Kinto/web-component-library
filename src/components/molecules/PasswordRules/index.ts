import PasswordRules from './PasswordRules.component';

export default PasswordRules;
