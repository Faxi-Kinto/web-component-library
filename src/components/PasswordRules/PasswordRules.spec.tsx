import React from 'react';
import renderer from 'react-test-renderer';
import PasswordRules from './PasswordRules.component';

test('PasswordRules renders correctly', () => {
  const component = renderer.create(<PasswordRules password="" />);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
