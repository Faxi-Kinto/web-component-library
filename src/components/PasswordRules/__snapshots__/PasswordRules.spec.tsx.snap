// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordRules renders correctly 1`] = `
.c0 {
  width: 100%;
}

.c0 > * {
  margin: 0 0 0.625rem;
}

.c0 > :last-child {
  margin: 0;
}

.c0 .password_rules__rule {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

.c0 .password_rules__rule > * {
  margin: 0 0.625rem 0 0;
}

.c0 .password_rules__rule > :last-child {
  margin: 0;
}

.c0 .password_rules__rule__icon {
  margin-top: 0.125rem;
  width: 0.625rem;
  height: 0.625rem;
}

.c2 {
  padding: 0;
  margin: 0;
  font-size: 1rem;
  color: #4E606A;
  font-weight: 300;
  white-space: pre-wrap;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 0.875rem;
}

<div
  className="c0 password_rules"
>
  <div
    className="password_rules__rule"
  >
    <div
      className="password_rules__rule__icon"
    >
      <div
        className="c1"
        size="0.875rem"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 fa-1x "
          color="#EB4820"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div>
      <span
        className="c2"
        color="#4E606A"
        fontSize="16px"
      >
        Contains at least 6 characters
      </span>
    </div>
  </div>
  <div
    className="password_rules__rule"
  >
    <div
      className="password_rules__rule__icon"
    >
      <div
        className="c1"
        size="0.875rem"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 fa-1x "
          color="#EB4820"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <div
      className="password_rules_wrapper_password_rule_number--message"
    >
      <span
        className="c2"
        color="#4E606A"
        fontSize="16px"
      >
        Contains at least one number (0-9)
      </span>
    </div>
  </div>
  <div
    className="password_rules__rule"
  >
    <div
      className="password_rules__rule__icon"
    >
      <div
        className="c1"
        size="0.875rem"
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 fa-1x "
          color="#EB4820"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </div>
    </div>
    <span
      className="c2"
      color="#4E606A"
      fontSize="16px"
    >
      Contains both lower (a-z) and upper case letters (A-Z)
    </span>
  </div>
</div>
`;
